File: PlaywireSDKAppsObjC/ads/fullscreen/rewarded/RewardedInterstitialViewController.m
//
//  RewardedInterstitialViewController.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/05/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "RewardedInterstitialViewController.h"
#import <Playwire-Swift.h>

@interface RewardedInterstitialViewController () <PWFullScreenAdDelegate>
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (strong, nonatomic) PWRewardedInterstitial *rewardedInterstitial;
@end

@implementation RewardedInterstitialViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self loadRewardedInterstitial];
}

- (void)loadRewardedInterstitial {
    self.rewardedInterstitial = [[PWRewardedInterstitial alloc] initWithAdUnitName:self.adUnitName delegate:self];
    
    // Use `[[PWLoadParams new] withTargeting:]` to pass your custom targets to ad request.
    // PWLoadParams *params = [[PWLoadParams new] withTargeting:@{
    //   @"age": @"18-32",
    //   @"page": @"travel"
    // }];
    // [self.rewardedInterstitial loadWithParams:params];
    
    [self.rewardedInterstitial load];
    self.statusLabel.text = [NSString stringWithFormat: @"⏳ The rewarded interstitial \"%@\" is loading.", self.adUnitName];
}

- (void)showRewardedInterstitial {
    if (!self.rewardedInterstitial.isLoaded) {
        // Load rewarded interstitial one more time or notify a user about error
        return;
    }
    
    [self.rewardedInterstitial showFromViewController:self];
}

#pragma mark - PWFullScreenAdDelegate -

- (void)fullScreenAdDidLoad:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"✅ The rewarded interstitial \"%@\" is loaded.", self.adUnitName];
    [self showRewardedInterstitial];
}

- (void)fullScreenAdDidFailToLoad:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"❌ Failed to load the rewarded interstitial \"%@\".", self.adUnitName];
}


- (void)fullScreenAdDidFailToPresentFullScreenContent:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"❌ Failed to show the rewarded interstitial \"%@\".", self.adUnitName];
}

- (void)fullScreenAdWillPresentFullScreenContent:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdWillDismissFullScreenContent:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdDidDismissFullScreenContent:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdDidRecordClick:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdDidRecordImpression:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"👍 The rewarded interstitial \"%@\" was successfully shown.", self.adUnitName];
}

- (void)fullScreenAdDidUserEarn:(PWFullScreenAd * _Nonnull)ad type:(NSString * _Nonnull)type amount:(double)amount {
    // Handle a reward regarding your business objectives.
    self.statusLabel.text = [NSString stringWithFormat: @"🎉 The reward is earned.\n Type: %@\n Amount: %0.f.", type, amount];
}

@end

File: PlaywireSDKAppsSwift/ads/fullscreen/rewarded/RewardedInterstitialViewController.swift
//
//  RewardedInterstitialViewController.swift
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 12/20/22.
//  Copyright © 2022 Playwire. All rights reserved.
//

import Playwire
import UIKit

final class RewardedInterstitialViewController: UIViewController, AdUnitViewControllerType {
    
    @IBOutlet weak var statusLabel: UILabel!

    // The ad unit name, e.g. 'banner-320x50', 'interstitial-home', 'rewarded-coins', etc.
    var adUnitName: String!
    
    private var rewardedInterstitial: PWRewardedInterstitial?
    private var name: String { adUnitName ?? "NULL"}
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadRewardedInterstitial()
    }
    
    private func loadRewardedInterstitial() {
        rewardedInterstitial = PWRewardedInterstitial(adUnitName: adUnitName, viewController: self, delegate: self)
        
        // Use `PWLoadParams().withTargeting()` to pass your custom targets to ad request.
        // let params = PWLoadParams().withTargeting(
        //  [
        //    "age": "18-32",
        //    "page": "travel"
        //  ]
        // )
        // rewardedInterstitial?.load(params: params)
        
        rewardedInterstitial?.load()
        
        statusLabel.text = "⏳ The rewarded interstitial \"\(name)\" is loading."
    }
    
    private func showRewardedInterstitial() {
        guard rewardedInterstitial?.isLoaded == true else {
            // Load rewarded interstitial one more time or notify a user about error
            return
        }
        rewardedInterstitial?.show()
    }
}

// MARK: - PWFullScreenAdDelegate -

extension RewardedInterstitialViewController: PWFullScreenAdDelegate {
    func fullScreenAdDidLoad(_ ad: PWFullScreenAd) {
        statusLabel.text = "✅ The rewarded interstitial \"\(name)\" is loaded."
        showRewardedInterstitial()
    }
    
    func fullScreenAdDidFailToLoad(_ ad: PWFullScreenAd) {
        statusLabel.text = "❌ Failed to load the rewarded interstitial \"\(name)\"."
    }
    
    func fullScreenAdWillPresentFullScreenContent(_ ad: PWFullScreenAd) {
        
    }
    
    func fullScreenAdWillDismissFullScreenContent(_ ad: PWFullScreenAd) {

    }
    
    func fullScreenAdDidDismissFullScreenContent(_ ad: PWFullScreenAd) {
    }
    
    func fullScreenAdDidFailToPresentFullScreenContent(_ ad: PWFullScreenAd) {
        statusLabel.text = "❌ Failed to show the rewarded interstitial \"\(name)\"."        
    }
    
    func fullScreenAdDidRecordImpression(_ ad: PWFullScreenAd) {
        statusLabel.text = "👍 The rewarded interstitial \"\(name))\" was successfully shown."
    }
    
    func fullScreenAdDidRecordClick(_ ad: PWFullScreenAd) {
        
    }
    
    func fullScreenAdDidUserEarn(_ ad: PWFullScreenAd, type: String, amount: Double) {
        // Handle a reward regarding your business objectives.
        statusLabel.text =  "🎉 The reward is earned.\n Type: \(type)\n Amount: \(String(format: "%0.f", amount))."
    }
}
